aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu

```diff
@@ -0,0 +1,521 @@
+  while (totalZ > 0) {
+    dim3 blocks(totalZ > 65535 ? 65535 : totalZ, blocksH);
+    atomicadaptiveaveragegradinput<<<blocks, threads, 0, at::cuda::getCurrentCUDAStream()>>>(
+        gradInput_data, gradOutput_data,
+        isizeT, isizeH, isizeW,
+        osizeT, osizeH, osizeW,
+        offsetZ);
+
+    totalZ -= 65535;
+    offsetZ += 65535;
+    AT_CUDA_CHECK(cudaGetLastError());
+  }
+}
+
+// 5D tensor B x D x T x H x w
+
+void adaptive_avg_pool3d_out_cuda_template(
+    Tensor& output,
+    const Tensor& input_,
+    IntArrayRef& output_size) {
+  TensorArg output_arg{output, "output", 1};
+  TensorArg input_arg{input_, "input_", 2};
+
+  checkAllSameGPU("adaptive_avg_pool3d_cuda", {output_arg, input_arg});
+
+  for (int64_t i = 0; i < input_.ndimension(); i++) {
+    AT_CHECK(
+        input_.size(i) > 0,
+        "adaptive_avg_pool3d_cuda(): expected input to have non-empty spatial dimensions, "
+        "but input has sizes ", input_.sizes(),
+        " with dimension ", i, " being empty");
+  }
+
+  AT_CHECK(
+      (input_.ndimension() == 4 || input_.ndimension() == 5),
+      "non-empty 4D or 5D (batch mode) tensor expected for input");
+
+  // the jit sometimes passes output_size.size() == 1
+  AT_CHECK(
+      output_size.size() == 1 || output_size.size() == 3,
+      "adaptive_avg_pool3d: internal error: output_size.size() must be 1 or 3");
+
+  int64_t osizeT = output_size[0];
+  int64_t osizeH = output_size[1];
+  int64_t osizeW = output_size[2];
+
+  int64_t sizeD, isizeT, isizeH, isizeW;
+  int64_t istrideD, istrideT, istrideH, istrideW;
+  int64_t totalZ;
+
+  const Tensor& input = input_.ndimension() == 4 ? input_ : input_.contiguous();
+
+  if (input.ndimension() == 4) {
+    sizeD = input.size(0);
+    isizeT = input.size(1);
+    isizeH = input.size(2);
+    isizeW = input.size(3);
+
+    istrideD = input.stride(0);
+    istrideT = input.stride(1);
+    istrideH = input.stride(2);
+    istrideW = input.stride(3);
+
+    output.resize_({sizeD, osizeT, osizeH, osizeW});
+
+    totalZ = sizeD * osizeT;
+  } else {
+    int64_t sizeB = input.size(0);
+    sizeD = input.size(1);
+    isizeT = input.size(2);
+    isizeH = input.size(3);
+    isizeW = input.size(4);
+
+    istrideD = input.stride(1);
+    istrideT = input.stride(2);
+    istrideH = input.stride(3);
+    istrideW = input.stride(4);
+
+    output.resize_({sizeB, sizeD, osizeT, osizeH, osizeW});
+
+    totalZ = sizeB * sizeD * osizeT;
+  }
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      input.scalar_type(), "adaptive_avg_pool3d_cuda", [&] {
+        scalar_t* input_data = input.data<scalar_t>();
+        scalar_t* output_data = output.data<scalar_t>();
+
+        adaptiveaveragepool_loop(
+            input_data, output_data,
+            totalZ,
+            isizeT, isizeH, isizeW,
+            osizeT, osizeH, osizeW,
+            istrideD, istrideT, istrideH, istrideW);
+      });
+}
+
+void adaptive_avg_pool3d_backward_out_cuda_template(
```

Comment:
@skrah @xmnlab 
Since this CUDA backward test fails and I'm struggling to fix this, could you take a look?

best regards